src/FormModel.php

```diff
@@ -227,36 +243,21 @@ public function setAttribute(string $name, $value): void
     public function processValidationResult(ResultSet $resultSet): void
     {
         $this->clearErrors();
+        $this->validated = false;
+
         foreach ($resultSet as $attribute => $result) {
-            if ($result->isValid() === false) {
-                $this->addErrors([$attribute => $result->getErrors()]);
+            if ($result->isValid() === false && ($errors = $result->getErrors()) !== []) {
+                $this->attributes[$attribute]->setErrors($errors);
             }
         }
         $this->validated = true;
     }
 
-    public function addError(string $attribute, string $error): void
```

Comment:
Are you sure what doesn't exists use cases for add error to model from user code?

Reply:
It is a violation of encapsulation so I don't want to have this method.
But we can add it back when the use case will appear.

Reply:
This one is questionable. There are use cases for adding error without having validator. For example, if we handle a form via external API and don't want to call two methods of this API but still want to display errors in the form per field.